warehouse/accounts/tasks.py

```diff
@@ -15,11 +15,48 @@
 from sqlalchemy import func
 
 from warehouse import tasks
-from warehouse.accounts.models import Email, User
+from warehouse.accounts.models import (
+    Email,
+    TermsOfServiceEngagement,
+    User,
+    UserTermsOfServiceEngagement,
+)
+from warehouse.accounts.services import IUserService
+from warehouse.email import send_user_terms_of_service_updated
 from warehouse.metrics import IMetricsService
 from warehouse.packaging.models import Release
 
 
+@tasks.task(ignore_result=True, acks_late=True)
+def notify_users_of_tos_update(request):
+    user_service = request.find_service(IUserService, context=None)
+    already_notified_subquery = (
+        request.db.query(UserTermsOfServiceEngagement.user_id)
+        .filter(
+            UserTermsOfServiceEngagement.revision
+            == request.registry.settings.get("terms.revision")
+        )
+        .filter(
+            UserTermsOfServiceEngagement.engagement.in_(
+                [TermsOfServiceEngagement.Notified, TermsOfServiceEngagement.Agreed]
+            )
+        )
+        .subquery()
+    )
+    users_to_notify = (
+        request.db.query(User)
+        .outerjoin(Email)
+        .filter(Email.verified == True, Email.primary == True)  # noqa E711
+        .filter(User.id.not_in(already_notified_subquery))
```

Comment:
Why not invert the `engagement` filter above, wouldn't that let us drop the need for this?

Reply:
not sure I understand, the goal here is to pull 1000 user objects who have previously agreed to the current ToS, or already been notified.